### ISSUE TITLE
Coloring "unused" identifiers

### ISSUE THREAD

**Issue by pnwamk** [racket/drracket#100]
In the Typed Racket implementation we have some files whose primary purpose is to declare the core datatypes that are used throughout the type checker. These definitions are written in a little DSL that expands into struct definitions w/ automatic provides and some other useful features.

I've notice in DrRacket on a fresh Racket build that _a lot_ of the identifiers in these files are now highlighted in red when moused over, reporting in a tool-top "no bound occurrences". While it may be true that no other _user-written_ code contains bound occurrences of these identifiers, there is macro generated code which does (e.g. provides, default generic function implementations, etc) and I wonder if the analysis for bound occurrences shouldn't include this code as well (in our case at least, it's just silly to have all of these identifiers reported as unused when they're being used just as we intended: by code after expansion).

For an example, see the following file: https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/rep/type-rep.rkt#L202

**Comment by rfindler** [comment]
This boils down to a program like this one, I guess?

```
#lang racket
(define-syntax-rule
  (define/provide x e)
  (begin (define x e)
         (provide x)))
(define/provide abcdef 1)
```

where the essence is that the bound and the binding identifier in the fully expanded program appear to have the same source location to Check Syntax?

I think that Check Syntax could detect this situation specially and avoid the red. And possibly avoid the tooltip, but I'm not sure about that. For example, in this case:

```
#lang racket
(define-syntax-rule
  (define/provide/random-reference x y e)
  (begin (define x e)
         (provide x)
         y))
(define/provide/random-reference abcdef
  abcdef
  1)
```

I think we do want to count that as one reference.

So, at least at the moment, I think we want to keep the tooltip counts the same, but change the way the red is decided.

Something like: "if the identifier appears in two different places in the tree that traversals.rkt traverses, then it is not a candidate for a red background"

What do you think @dfeltey ?

**Comment by pnwamk** [comment]
Yes, that snippet has the same behavior I'm seeing (sorry for not minimizing the example).

The red highlighting is definitely more offensive than the tooltip.

**Comment by pnwamk** [comment]
Another small example where a function is highlighted in red when it shouldn't be (IMHO):

```
#lang racket

(provide (all-defined-out))
(define (id x) x)
```

`id` is highlighted with red when moused over in this example.

This may just be the same issue requiring the same adjustment/fix.

**Comment by dfeltey** [comment]
I've pushed a change that should remove some of the excessive red coloring on identifiers, and it appears to do so for the Typed Racket file you point out. 

I don't have a fix for the `(all-defined-out)` issue yet though, the problem there is that the identifiers provided in its expansion aren't `original?` so check-syntax just ignores them right now.

**Comment by soegaard** [comment]
It seems this issue has been resolved:

<img width="442" alt="Image" src="https://github.com/user-attachments/assets/b19eef7c-bea8-4e45-b44c-910507daef0f" />

<img width="369" alt="Image" src="https://github.com/user-attachments/assets/656b62e2-767f-4093-8b9b-e5e9a397dba2" />

Time to close?
